spirv_glsl.hpp

```diff
@@ -399,6 +399,8 @@ class CompilerGLSL : public Compiler
 	void end_scope_decl(const std::string &decl);
 
 	Options options;
+	bool declaredKHR_shader_subgroup_basic = false;
+	bool declaredKHR_shader_subgroup_vote = false;
```

Comment:
ballot is missing.

Reply:
I'd rather track the following (instead of KHR extensions):
- needs group voting
- needs subgroup masks
- needs subgroup elect
- needs subgroup ballots

@HansKristian-Work whats the ordering between `emit_header` and `emit_subgroup_op`, the logical one?

Reply:
Not sure I understand the question.

Reply:
will `emit_header` be called before `emit_subgroup_op` ?

Reply:
> I'd rather track the following (instead of KHR extensions):
> 
> * needs subgroup voting on booleans
> * needs subgroup voting on non-booleans
> * needs subgroup masks
> * needs subgroup elect
> * needs subgroup ballots
> 

I'd also track:
- needs the input `gl_SubgroupID` or `gl_NumSubgroups` (requires NV)
- needs the input `gl_SubgroupSize` requires (requires NV, AMD or ARB_shader_ballot )
- needs the input `gl_SubgroupInvocationID` (requires NV or ARB_shader_ballot )

Basically we need to analyse the SPIR-V opcodes and not the declared extension usage, because outside the NV ecosystem the support for the Vulkan extension is "partial".

Then we can establish which combos of extensions can potentially implement the shader.

Reply:
Spirv-cross kind of already does it by requiring specific extensions when found specific feature (builtin input/opcode/etc..), so I think it's better to use exisiting code and leave those `declared_<extension_name>` flags

for example see https://github.com/KhronosGroup/SPIRV-Cross/blob/master/spirv_glsl.cpp#L7356 or https://github.com/KhronosGroup/SPIRV-Cross/blob/master/spirv_glsl.cpp#L10906

Reply:
yes, but...

these are part of `builtin_to_glsl` and `emit_instruction` which get called after the header is already emitted?

Anyway @HansKristian-Work suggested we use `traverse_all_reachable_opcodes`

Reply:
hm, yea, you're right... No idea why spirv-cross does it then..

Reply:
Ah, because it then forces recompilation. So actually it will do

Reply:
> will `emit_header` be called before `emit_subgroup_op` ?

emit_header will be called before any functions. emit_header and emit_resources are the first things to be emitted.

Reply:
i figured `require_extension_internal` forces recompilation so it doesnt really matter

Reply:
> i figured `require_extension_internal` forces recompilation so it doesnt really matter

sounds like compile times will explode, its best to run poor-man's static analysis with `traverse_all_reachable_opcodes` before `emit_header` and tag all the opcodes groups and builtin groups that are of interest.

Reply:
https://github.com/KhronosGroup/SPIRV-Cross/pull/1476/files#r498956448